File: templates/zig/src/plugins/agno.zig
//! Agno → Borgkit Plugin (Zig) — HTTP Bridge
//!
//! Wraps a deployed Agno FastAPI agent server so it participates in the
//! Borgkit mesh as a standard IAgent.  Uses `std.http.Client` for all HTTP
//! communication — no external dependencies required.
//!
//! ── Agno API contract ─────────────────────────────────────────────────────────
//!
//!   POST {base_url}/run                          (default, single-agent)
//!     Body:   { "message": "...", "stream": false }
//!     Response: { "content": "...", "messages": [...] }
//!
//!   POST {base_url}/v1/agents/{agent_id}/runs    (when agent_id is set)
//!     Same body / response shape.
//!
//! The plugin tries `content` first in the response; if absent it falls back
//! to the `content` field of the last entry in the `messages` array.
//!
//! ── Setup ──────────────────────────────────────────────────────────────────────
//!
//!   Start an Agno FastAPI server:
//!     from agno.agent import Agent
//!     from agno.playground import Playground
//!     agent = Agent(...)
//!     app   = Playground(agents=[agent]).get_app()
//!     # uvicorn app:app --port 7777
//!
//! ── Usage ──────────────────────────────────────────────────────────────────────
//!
//!   const agno = @import("plugins/agno.zig");
//!
//!   var service = agno.AgnoService{
//!       .base_url  = "http://localhost:7777",
//!       .agent_id  = "my_agent",   // optional
//!   };
//!   var plugin = agno.AgnoPlugin.init(allocator);
//!   defer plugin.deinit();
//!
//!   const Wrapped = wrapped_agent.WrappedAgent(agno.AgnoService, agno.AgnoPlugin);
//!   var agent = Wrapped.init(&service, &plugin, .{
//!       .agent_id = "borgkit://agent/agno",
//!       .owner    = "0xYourWallet",
//!   }, allocator);
//!
//!   const resp = agent.handleRequest(req);

const std     = @import("std");
const types   = @import("../types.zig");
const iplugin = @import("iPlugin.zig");

// ── Service config (the "agent" token) ────────────────────────────────────────

/// Configuration for an Agno FastAPI HTTP endpoint.
pub const AgnoService = struct {
    /// Base URL, e.g. "http://localhost:7777".
    base_url: []const u8 = "http://localhost:7777",

    /// POST path used when agent_id is not set (default: "/run").
    invoke_route: []const u8 = "/run",

    /// Optional Agno agent ID.  When set the plugin posts to
    /// `{base_url}/v1/agents/{agent_id}/runs` instead of `{base_url}{invoke_route}`.
    agent_id: ?[]const u8 = null,

    /// Optional API key sent as "Authorization: Bearer <key>".
    api_key: ?[]const u8 = null,
};

// ── Plugin ────────────────────────────────────────────────────────────────────

/// HTTP bridge plugin for Agno FastAPI agent servers.
///
/// Stores a reusable `std.http.Client`; call `deinit()` when done.
pub const AgnoPlugin = struct {
    allocator: std.mem.Allocator,
    client:    std.http.Client,

    pub fn init(allocator: std.mem.Allocator) AgnoPlugin {
        return .{
            .allocator = allocator,
            .client    = std.http.Client{ .allocator = allocator },
        };
    }

    pub fn deinit(self: *AgnoPlugin) void {
        self.client.deinit();
    }

    // ── IPlugin interface ───────────────────────────────────────────────────

    /// Return a single "invoke" capability — capabilities are static for HTTP
    /// bridge plugins.  Pass explicit capabilities at wrap time via
    /// `WrappedAgent.init` if you need multiple.
    pub fn extractCapabilities(
        _self:     *AgnoPlugin,
        _agent:    *AgnoService,
        allocator: std.mem.Allocator,
    ) []iplugin.CapabilityDescriptor {
        const caps = allocator.alloc(iplugin.CapabilityDescriptor, 1) catch
            return &[_]iplugin.CapabilityDescriptor{};
        caps[0] = .{
            .name        = "invoke",
            .description = "Invoke the Agno agent via its FastAPI HTTP server",
        };
        return caps;
    }

    /// Build the Agno run JSON body:
    ///   { "message": <payload>, "stream": false }
    ///
    /// Extracts `message`, `input`, or `query` key from the request payload if
    /// present; otherwise uses the raw payload string as the message.
    pub fn translateRequest(
        _self:     *AgnoPlugin,
        req:       types.AgentRequest,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        // Try to extract a string content field from the JSON payload
        const content = blk: {
            const parsed = std.json.parseFromSlice(
                std.json.Value, allocator, req.payload, .{ .ignore_unknown_fields = true },
            ) catch break :blk req.payload;
            defer parsed.deinit();

            if (parsed.value == .object) {
                for ([_][]const u8{ "message", "input", "query" }) |key| {
                    if (parsed.value.object.get(key)) |v| {
                        if (v == .string) {
                            const copy = try allocator.dupe(u8, v.string);
                            break :blk copy;
                        }
                    }
                }
            }
            break :blk try allocator.dupe(u8, req.payload);
        };
        defer allocator.free(content);

        // JSON-encode the content string so it is safe inside the body
        var escaped = std.ArrayList(u8).init(allocator);
        defer escaped.deinit();
        try std.json.encodeJsonString(content, .{}, escaped.writer());

        return std.fmt.allocPrint(allocator,
            \\{{"message":{s},"stream":false}}
            , .{escaped.items});
    }

    /// Extract the agent reply from the Agno response.
    ///
    /// Tries `content` at the top level first; if absent or empty, walks
    /// `messages` in reverse and returns the first non-empty `content` value
    /// found.  Falls back to the raw JSON when nothing matches.
    pub fn translateResponse(
        _self:      *AgnoPlugin,
        request_id: []const u8,
        raw:        []const u8,
    ) types.AgentResponse {
        const content = extractAgnoContent(raw) orelse raw;
        return types.AgentResponse.success(request_id, content);
    }

    /// POST the translated body to the appropriate Agno URL and return the
    /// raw response bytes.
    ///
    /// URL selection:
    ///   • agent_id set → `{base_url}/v1/agents/{agent_id}/runs`
    ///   • otherwise    → `{base_url}{invoke_route}`
    pub fn invokeNative(
        self:      *AgnoPlugin,
        agent:     *AgnoService,
        input:     []const u8,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        const url = if (agent.agent_id) |id|
            try std.fmt.allocPrint(
                allocator, "{s}/v1/agents/{s}/runs",
                .{ trimSlash(agent.base_url), id },
            )
        else
            try std.fmt.allocPrint(
                allocator, "{s}{s}",
                .{ trimSlash(agent.base_url), agent.invoke_route },
            );
        defer allocator.free(url);

        return self.postJson(agent, url, input, allocator);
    }

    // ── Internal HTTP helper ────────────────────────────────────────────────

    fn postJson(
        self:      *AgnoPlugin,
        agent:     *AgnoService,
        url:       []const u8,
        payload:   []const u8,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        var auth_buf: [512]u8 = undefined;
        var hdrs_buf: [3]std.http.Header = undefined;
        var n: usize = 0;
        hdrs_buf[n] = .{ .name = "Content-Type", .value = "application/json" };
        n += 1;
        hdrs_buf[n] = .{ .name = "Accept", .value = "application/json" };
        n += 1;
        if (agent.api_key) |key| {
            const bearer = try std.fmt.bufPrint(&auth_buf, "Bearer {s}", .{key});
            hdrs_buf[n] = .{ .name = "Authorization", .value = bearer };
            n += 1;
        }
        const extra = hdrs_buf[0..n];

        var body = std.ArrayList(u8).init(allocator);
        errdefer body.deinit();

        const fr = try self.client.fetch(.{
            .method           = .POST,
            .location         = .{ .url = url },
            .extra_headers    = extra,
            .payload          = payload,
            .response_storage = .{ .dynamic = &body },
        });

        const code = @intFromEnum(fr.status);
        if (code < 200 or code >= 300) return error.AgnoBadStatus;

        return body.toOwnedSlice();
    }
};

// ── Helpers ───────────────────────────────────────────────────────────────────

/// Try to extract a meaningful text reply from an Agno response JSON.
///
/// Strategy:
///   1. Return `content` at the top level if it is a non-empty string.
///   2. Walk `messages` in reverse; return the first non-empty `content`.
///   3. Return null so the caller falls back to raw bytes.
fn extractAgnoContent(raw: []const u8) ?[]const u8 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const parsed = std.json.parseFromSlice(
        std.json.Value, alloc, raw, .{ .ignore_unknown_fields = true },
    ) catch return null;

    const root = switch (parsed.value) {
        .object => |obj| obj,
        else    => return null,
    };

    // 1. Top-level "content" field
    if (root.get("content")) |v| {
        if (v == .string and v.string.len > 0) return v.string;
    }

    // 2. Last non-empty content in messages array
    const messages_val = root.get("messages") orelse return null;
    const arr = switch (messages_val) {
        .array => |a| a.items,
        else   => return null,
    };

    var i = arr.len;
    while (i > 0) {
        i -= 1;
        const msg = arr[i];
        if (msg != .object) continue;
        const content = msg.object.get("content") orelse continue;
        if (content == .string and content.string.len > 0) return content.string;
    }

    return null;
}

fn trimSlash(s: []const u8) []const u8 {
    var end = s.len;
    while (end > 0 and s[end - 1] == '/') end -= 1;
    return s[0..end];
}

File: templates/rust/src/plugins/agno.rs
//! Agno → Borgkit Plugin (Rust) — HTTP Bridge
//!
//! Wraps a deployed Agno FastAPI server so it participates in the Borgkit mesh
//! as a standard `IAgent`.  Agno agents are served via `app.run()` which starts
//! a FastAPI server; this plugin bridges both the multi-agent route and the
//! simpler single-agent `/run` route.
//!
//! ── API contract ──────────────────────────────────────────────────────────────
//!
//!   Multi-agent route (agent_id is set):
//!     POST {base_url}/v1/agents/{agent_id}/runs
//!       Body:     { "message": "...", "stream": false }
//!       Response: { "content": "...", "run_id": "..." }
//!
//!   Single-agent route (agent_id is None):
//!     POST {base_url}/run
//!       Body:     { "message": "...", "stream": false }
//!       Response: { "content": "...", "messages": [...] }
//!
//!   Streaming (stream: true) — returns newline-delimited JSON events;
//!   this plugin always disables streaming and reads the full response.
//!
//! ── Setup ─────────────────────────────────────────────────────────────────────
//!
//!   Start your Agno agent server with:
//!     from agno.app import App
//!     app = App(agents=[my_agent])
//!     app.run()               # starts on http://localhost:7777 by default
//!
//! ── Usage ─────────────────────────────────────────────────────────────────────
//!
//!   use borgkit::plugins::agno::{AgnoPlugin, AgnoService};
//!   use borgkit::plugins::base::PluginConfig;
//!
//!   let service = AgnoService {
//!       base_url:     "http://localhost:7777".to_string(),
//!       agent_id:     Some("my-research-agent".to_string()),
//!       capabilities: vec![("research".to_string(), "Research a topic".to_string())],
//!       ..Default::default()
//!   };
//!
//!   let plugin = AgnoPlugin::with_timeout(120);
//!   let agent  = plugin.wrap(service, PluginConfig {
//!       agent_id:     "borgkit://agent/agno-researcher".to_string(),
//!       owner:        "0xYourWallet".to_string(),
//!       network_host: "localhost".to_string(),
//!       network_port: 6174,
//!       ..Default::default()
//!   });

use async_trait::async_trait;
use serde_json::{json, Value};

use crate::plugins::base::{CapabilityDescriptor, BorgkitPlugin};
use crate::request::AgentRequest;
use crate::response::AgentResponse;

// ── Service config (the "agent" token) ───────────────────────────────────────

/// Configuration for an Agno FastAPI HTTP endpoint.
pub struct AgnoService {
    /// Base URL, e.g. `http://localhost:7777`.
    pub base_url: String,

    /// Optional Agno agent identifier.
    ///
    /// When set, requests are routed to `/v1/agents/{agent_id}/runs`.
    /// When `None`, requests are sent to `invoke_route` (`"/run"` by default).
    pub agent_id: Option<String>,

    /// Fallback POST path used when `agent_id` is `None` (default: `"/run"`).
    pub invoke_route: String,

    /// Optional Bearer token sent as `Authorization: Bearer <api_key>`.
    pub api_key: Option<String>,

    /// Whether to request a streaming response from the server (default: `false`).
    ///
    /// This plugin always reads the full JSON response regardless of this setting;
    /// it is forwarded in the request body for server-side stream control.
    pub stream: bool,

    /// Explicit capability list: `(name, description)` pairs.
    ///
    /// When empty a single `"run"` capability is synthesised automatically.
    pub capabilities: Vec<(String, String)>,
}

impl Default for AgnoService {
    fn default() -> Self {
        Self {
            base_url:     "http://localhost:7777".to_string(),
            agent_id:     None,
            invoke_route: "/run".to_string(),
            api_key:      None,
            stream:       false,
            capabilities: vec![],
        }
    }
}

// ── Plugin ────────────────────────────────────────────────────────────────────

/// HTTP bridge plugin for Agno FastAPI servers.
pub struct AgnoPlugin {
    client: reqwest::Client,
}

impl AgnoPlugin {
    /// Create with a default 60-second timeout.
    pub fn new() -> Self {
        Self::with_timeout(60)
    }

    /// Create with a custom per-request timeout in seconds.
    pub fn with_timeout(secs: u64) -> Self {
        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(secs))
            .build()
            .unwrap_or_else(|_| reqwest::Client::new());
        Self { client }
    }
}

impl Default for AgnoPlugin {
    fn default() -> Self { Self::new() }
}

// ── BorgkitPlugin impl ────────────────────────────────────────────────────────

#[async_trait]
impl BorgkitPlugin<AgnoService> for AgnoPlugin {
    fn extract_capabilities(&self, service: &AgnoService) -> Vec<CapabilityDescriptor> {
        if service.capabilities.is_empty() {
            return vec![CapabilityDescriptor {
                name:           "run".to_string(),
                description:    "Run the Agno agent with a message".to_string(),
                input_schema:   None,
                output_schema:  None,
                price_per_call: None,
            }];
        }
        service.capabilities.iter().map(|(name, desc)| CapabilityDescriptor {
            name:           name.clone(),
            description:    desc.clone(),
            input_schema:   None,
            output_schema:  None,
            price_per_call: None,
        }).collect()
    }

    /// Build the Agno request body from an `AgentRequest`.
    ///
    /// The payload field `"message"`, `"input"`, or `"query"` (first found) becomes
    /// the message content; otherwise the whole payload is stringified.
    /// Streaming is always disabled (`"stream": false`).
    fn translate_request(&self, request: &AgentRequest) -> Result<Value, String> {
        let message = request.payload.get("message")
            .or_else(|| request.payload.get("input"))
            .or_else(|| request.payload.get("query"))
            .map(|v| if v.is_string() {
                v.as_str().unwrap_or("").to_string()
            } else {
                v.to_string()
            })
            .unwrap_or_else(|| request.payload.to_string());

        Ok(json!({
            "message": message,
            "stream":  false,
        }))
    }

    /// Extract the reply from the Agno response.
    ///
    /// Tries `content` first; if absent, walks `messages` in reverse and returns
    /// the first assistant-role message content found.
    fn translate_response(&self, request_id: &str, output: Value) -> AgentResponse {
        let content = output.get("content")
            .and_then(|v| if v.is_string() { v.as_str().map(|s| s.to_string()) } else { None })
            .or_else(|| {
                output.get("messages")
                    .and_then(|m| m.as_array())
                    .and_then(|msgs| {
                        msgs.iter().rev().find_map(|msg| {
                            let role = msg.get("role").and_then(|r| r.as_str()).unwrap_or("");
                            if role == "assistant" || role == "agent" {
                                msg.get("content").and_then(|c| {
                                    if c.is_string() {
                                        c.as_str().map(|s| s.to_string())
                                    } else {
                                        Some(c.to_string())
                                    }
                                })
                            } else {
                                None
                            }
                        })
                    })
            })
            .unwrap_or_else(|| output.to_string());

        AgentResponse::success(
            request_id.to_string(),
            json!({ "content": content, "raw": output }),
        )
    }

    /// POST the translated body to the appropriate Agno endpoint.
    ///
    /// When `agent_id` is set the route is `/v1/agents/{agent_id}/runs`;
    /// otherwise `invoke_route` is used.  An optional Bearer token is added
    /// when `api_key` is configured.
    async fn invoke_native(
        &self,
        service: &AgnoService,
        input:   Value,
    ) -> Result<Value, String> {
        let route = match &service.agent_id {
            Some(id) => format!("/v1/agents/{}/runs", id),
            None     => service.invoke_route.clone(),
        };

        let url = format!("{}{}", service.base_url.trim_end_matches('/'), route);

        let mut req = self.client.post(&url).json(&input);

        if let Some(ref key) = service.api_key {
            req = req.header("Authorization", format!("Bearer {key}"));
        }

        req.send()
            .await
            .map_err(|e| format!("Agno HTTP error: {e}"))?
            .json::<Value>()
            .await
            .map_err(|e| format!("Agno response parse error: {e}"))
    }
}
